### PR TITLE
enh: Handle BinOp in ArrayItem Reference

### LINKED ISSUE #5687: (PRIMA) Handle BinOp in ArrayItem Reference
MRE : 
```fortran 
program test
    integer :: X(2)
    X = [0,1]
    X(X + 1) = 1
    print *, X
end program
```

LFortran Error : 
```console
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  Binary file "/mnt/c/lfortran/src/bin/lfortran", in _start
  File "/build/glibc-LcI20x/glibc-2.31/csu/../csu/libc-start.c", line 308, in __libc_start_main
  File "/mnt/c/lfortran/src/bin/lfortran.cpp", line 2732, in main
    return main_app(argc, argv);
  File "/mnt/c/lfortran/src/bin/lfortran.cpp", line 2670, in main_app(int, char**)
    err = compile_src_to_object_file(arg_file, tmp_o, false,
  File "/mnt/c/lfortran/src/bin/lfortran.cpp", line 1049, in (anonymous namespace)::compile_src_to_object_file(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, LCompilers::CompilerOptions&, LCompilers::PassManager&)
    res = fe.get_llvm3(*asr, lpm, diagnostics, infile);
  File "/mnt/c/lfortran/src/lfortran/fortran_evaluator.cpp", line 382, in LCompilers::FortranEvaluator::get_llvm3(LCompilers::ASR::TranslationUnit_t&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    compiler_options, run_fn, infile);
  File "/mnt/c/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 10412, in LCompilers::asr_to_llvm(LCompilers::ASR::TranslationUnit_t&, LCompilers::diag::Diagnostics&, llvm::LLVMContext&, Allocator&, LCompilers::PassManager&, LCompilers::CompilerOptions&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    v.visit_asr((ASR::asr_t&)asr);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr_base_visitor.h", line 316, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_asr(LCompilers::ASR::asr_t const&)
    void visit_asr(const asr_t &b) { visit_asr_t(b, self()); }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr_base_visitor.h", line 292, in void LCompilers::ASR::visit_asr_t<LCompilers::ASRToLLVMVisitor>(LCompilers::ASR::asr_t const&, LCompilers::ASRToLLVMVisitor&)
    case asrType::unit: { v.visit_unit((const unit_t &)x); return; }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr_base_visitor.h", line 317, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_unit(LCompilers::ASR::unit_t const&)
    void visit_unit(const unit_t &b) { visit_unit_t(b, self()); }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr_base_visitor.h", line 22, in void LCompilers::ASR::visit_unit_t<LCompilers::ASRToLLVMVisitor>(LCompilers::ASR::unit_t const&, LCompilers::ASRToLLVMVisitor&)
    case unitType::TranslationUnit: { v.visit_TranslationUnit((const TranslationUnit_t &)x); return; }
  File "/mnt/c/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 988, in LCompilers::ASRToLLVMVisitor::visit_TranslationUnit(LCompilers::ASR::TranslationUnit_t const&)
    visit_symbol(*item.second);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr_base_visitor.h", line 319, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_symbol(LCompilers::ASR::symbol_t const&)
    void visit_symbol(const symbol_t &b) { visit_symbol_t(b, self()); }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr_base_visitor.h", line 30, in void LCompilers::ASR::visit_symbol_t<LCompilers::ASRToLLVMVisitor>(LCompilers::ASR::symbol_t const&, LCompilers::ASRToLLVMVisitor&)
    case symbolType::Program: { v.visit_Program((const Program_t &)x); return; }
  File "/mnt/c/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 3292, in LCompilers::ASRToLLVMVisitor::visit_Program(LCompilers::ASR::Program_t const&)
    this->visit_stmt(*x.m_body[i]);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr_base_visitor.h", line 336, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_stmt(LCompilers::ASR::stmt_t const&)
    void visit_stmt(const stmt_t &b) { visit_stmt_t(b, self()); }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr_base_visitor.h", line 85, in void LCompilers::ASR::visit_stmt_t<LCompilers::ASRToLLVMVisitor>(LCompilers::ASR::stmt_t const&, LCompilers::ASRToLLVMVisitor&)
    case stmtType::WhileLoop: { v.visit_WhileLoop((const WhileLoop_t &)x); return; }
  File "/mnt/c/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 6084, in LCompilers::ASRToLLVMVisitor::visit_WhileLoop(LCompilers::ASR::WhileLoop_t const&)
    create_loop(x.m_name, [=]() {
  File "/mnt/c/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 321, in void LCompilers::ASRToLLVMVisitor::create_loop<LCompilers::ASRToLLVMVisitor::visit_WhileLoop(LCompilers::ASR::WhileLoop_t const&)::'lambda'(), LCompilers::ASRToLLVMVisitor::visit_WhileLoop(LCompilers::ASR::WhileLoop_t const&)::'lambda0'()>(char*, LCompilers::ASRToLLVMVisitor::visit_WhileLoop(LCompilers::ASR::WhileLoop_t const&)::'lambda'(), LCompilers::ASRToLLVMVisitor::visit_WhileLoop(LCompilers::ASR::WhileLoop_t const&)::'lambda0'())
    loop_body();
  File "/mnt/c/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 6090, in LCompilers::ASRToLLVMVisitor::visit_WhileLoop(LCompilers::ASR::WhileLoop_t const&)::'lambda0'()::operator()() const
    this->visit_stmt(*x.m_body[i]);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr_base_visitor.h", line 336, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_stmt(LCompilers::ASR::stmt_t const&)
    void visit_stmt(const stmt_t &b) { visit_stmt_t(b, self()); }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr_base_visitor.h", line 56, in void LCompilers::ASR::visit_stmt_t<LCompilers::ASRToLLVMVisitor>(LCompilers::ASR::stmt_t const&, LCompilers::ASRToLLVMVisitor&)
    case stmtType::Assignment: { v.visit_Assignment((const Assignment_t &)x); return; }
  File "/mnt/c/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 5029, in LCompilers::ASRToLLVMVisitor::visit_Assignment(LCompilers::ASR::Assignment_t const&)
    this->visit_expr(*x.m_target);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr_base_visitor.h", line 384, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_expr(LCompilers::ASR::expr_t const&)
    void visit_expr(const expr_t &b) { visit_expr_t(b, self()); }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr_base_visitor.h", line 173, in void LCompilers::ASR::visit_expr_t<LCompilers::ASRToLLVMVisitor>(LCompilers::ASR::expr_t const&, LCompilers::ASRToLLVMVisitor&)
    case exprType::ArrayItem: { v.visit_ArrayItem((const ArrayItem_t &)x); return; }
  File "/mnt/c/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 2365, in LCompilers::ASRToLLVMVisitor::visit_ArrayItem(LCompilers::ASR::ArrayItem_t const&)
    this->visit_expr_wrapper(curr_idx.m_right, true);
  File "/mnt/c/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 5550, in LCompilers::ASRToLLVMVisitor::visit_expr_wrapper(LCompilers::ASR::expr_t*, bool)
    this->visit_expr(*x);
  File "/mnt/c/lfortran/src/libasr/../libasr/asr_base_visitor.h", line 384, in LCompilers::ASR::BaseVisitor<LCompilers::ASRToLLVMVisitor>::visit_expr(LCompilers::ASR::expr_t const&)
    void visit_expr(const expr_t &b) { visit_expr_t(b, self()); }
  File "/mnt/c/lfortran/src/libasr/../libasr/asr_base_visitor.h", line 173, in void LCompilers::ASR::visit_expr_t<LCompilers::ASRToLLVMVisitor>(LCompilers::ASR::expr_t const&, LCompilers::ASRToLLVMVisitor&)
    case exprType::ArrayItem: { v.visit_ArrayItem((const ArrayItem_t &)x); return; }
  File "/mnt/c/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 2374, in LCompilers::ASRToLLVMVisitor::visit_ArrayItem(LCompilers::ASR::ArrayItem_t const&)
    bool is_bindc_array = ASRUtils::expr_abi(x.m_v) == ASR::abiType::BindC;
LCompilersException: Cannot extract the ABI of i32[2] expression.
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #5687

### CODE REVIEW

**Comment by AnshMehta1** [comment]
I think , things are breaking in `array_op_simplifier` . I think using `NO_EXPERIMENTAL_SIMPLIFIER` would work.

File: integration_tests/CMakeLists.txt [hunk 1]
```diff
@@ -540,6 +540,7 @@ RUN(NAME arrays_59 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
 RUN(NAME arrays_60 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
 RUN(NAME arrays_61 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran)
 RUN(NAME arrays_62 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran)
+RUN(NAME arrays_64 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran)
```
Comment (Pranavchiku) [inline integration_tests/CMakeLists.txt:543]:
```suggestion
RUN(NAME arrays_64 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran NO_EXPERIMENTAL_SIMPLIFIER)
```

**Comment by gxyd** [comment]
Sorry to bug here, @Pranavchiku , I think @certik isn't really happy with the use of label `NO_EXPERIMENTAL_SIMPLIFIER` anymore, as the push is to finally get the *simplifier pass* as default, e.g.; message from yesterday: https://lfortran.zulipchat.com/#narrow/channel/452774-Internal/topic/Simplifier.20pass/near/489092682, just and FYI in-case you missed it.

**Comment by certik** [comment]
Resolve the conflict and merge it.

**Pull request merged**